crates/shared/src/event_handling.rs

```diff
@@ -69,51 +71,143 @@ where
-        // large block ranges.
-        let last_handled_block = match self.last_handled_block {
-            Some(block) => block,
-            None => self.store.last_event_block().await?,
+    /// Defines block range, for which events should be fetched
+    /// Returns 3 parameters:
+    /// 1. Optional block number range for fetching reorg safe history (if it's needed)
+    /// 2. List of block numbers with hashes for fetching reorg unsafe blocks
+    /// 3. Bool indicator if reorg happened for reorg unsafe blocks
+    async fn event_block_range(
+        &self,
+    ) -> Result<(Option<RangeInclusive<u64>>, Vec<BlockNumberHash>, bool)> {
+        let handled_blocks = if self.last_handled_blocks.is_empty() {
+            let last_handled_block = self.store.last_event_block().await?;
+            self.block_retriever
+                .blocks(RangeInclusive::new(last_handled_block, last_handled_block))
+                .await?
+        } else {
+            self.last_handled_blocks.clone()
         };
-        let current_block = self.block_retriever.current_block_number().await?;
-        let from_block = last_handled_block.saturating_sub(MAX_REORG_BLOCK_COUNT);
-        anyhow::ensure!(
-            from_block <= current_block,
+
+        let current_block = self.block_retriever.current_block().await?;
+        let current_block_number = current_block.number.context("missing number")?.as_u64();
+        let current_block_hash = current_block.hash.context("missing hash")?;
+        let (last_handled_block_number, last_handled_block_hash) = *handled_blocks.last().unwrap();
+        tracing::debug!(
+            "current block: {} - {:?}, handled_blocks: {:?}",
+            current_block_number,
+            current_block_hash,
+            handled_blocks,
+        );
+
+        // handle special case which happens most of the time (no reorg, just one new block is added)
+        if current_block.parent_hash == last_handled_block_hash {
+            return Ok((
+                None,                                             // history range
+                vec![(current_block_number, current_block_hash)], // latest blocks
+                false,                                            // is reorg
+            ));
+        }
+
+        // handle special case when no new block is added
+        if (current_block_number, current_block_hash)
+            == (last_handled_block_number, last_handled_block_hash)
+        {
+            return Ok((
+                None,   // history range
+                vec![], // latest blocks
+                false,  // is reorg
+            ));
+        }
+
+        // full range of blocks which are considered for event update
+        let range = RangeInclusive::new(
+            last_handled_block_number.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            current_block_number,
+        );
+        ensure!(
+            !range.is_empty(),
             format!(
```

Comment:
I think `ensure` automatically `format`s - so you should be able to remove it and:
```rust
ensure!(
    condition,
    "string with {} - the answer is {}",
    "formatting",
    42;
);
```

Reply:
Yes, missed this since it was copy paste from old code.